Snapshot overlay driver metastore

dmcgowan opened the pull request:
Provides an interface for storing snapshot driver metadata in consistent way across driver implementations. Allows driver implementations to differentiate based on their snapshot storage, avoiding duplication of effort on storing metadata about the state and relationships of those snapshots.

Includes an implementation using https://github.com/cznic/ql. An implementation using files similar to the existing overlay driver is in progress, and an implementation using boltdb is planned.

dmcgowan commented:
Got btrfs driver updated, can we get this merged so I don't have to combine the PRs.

samuelkarp commented:
Before we merge this, I'm curious to understand your reasoning behind picking https://github.com/cznic/ql over something a bit more traditional like sqlite.  (I'm not opposed to any particular implementation, I just hadn't heard any discussion on the choice before and would like to understand the pros/cons behind it.)

dmcgowan commented:
@samuelkarp more of just a starter, going to do a few different implementation of the interface to see which is best, boltdb is next.

samuelkarp commented:
@dmcgowan Got it, so the intention is to have multiple implementations of the metastore?

stevvooe commented:
@samuelkarp `sqlite` is great but due to Go's calling conventions there is an extra expense and complexity of deployment using it verses an embedded solution.

Storage-wise, we are realistically considering `boltdb` and `cznic/ql`, as they are both performant, robust and stable implementations. We have experience with `boltdb`, from swarmkit and docker, and `cznic/ql` was suggested by @philips after they used it in `rkt`.

samuelkarp commented:
@stevvooe (maybe we should move this to a separate issue?) Is the overhead in cgo enough to make sqlite undesirable even though it's an extremely well-tested (both test coverage and in production settings) datastore?

stevvooe commented:
> @stevvooe (maybe we should move this to a separate issue?) Is the overhead in cgo enough to make sqlite undesirable even though it's an extremely well-tested (both test coverage and in production settings) datastore?

I don't think it needs a full issue for discussion.

You are also not considering the complexity of using and deploying sqlite with Go.

samuelkarp commented:
> You are also not considering the complexity of using and deploying sqlite with Go.

Great!  What should I consider here?  I know that sqlite uses a [very minimal subset of the C standard library](https://www.sqlite.org/selfcontained.html) and provides an [amalgamation that makes it easy to compile](https://www.sqlite.org/howtocompile.html).  I know that there is a [go-sqlite3 library](https://github.com/mattn/go-sqlite3) to make it usable as a `database/sql`.  I know that sqlite is [tested rigorously](https://www.sqlite.org/testing.html) and is used in a large corpus of production software.

I also know that we don't expect individual users to deploy containerd themselves, but that containerd is meant to be packaged alongside higher-level software that uses it (like Docker).  Producing an installable package for the higher-level software + containerd can be a responsibility of the maintainer of that higher-level software.

And to reiterate: I'm not saying that we *should* use sqlite or *should not* use `boltdb` or `cznic/ql`.  I'm just trying to understand what benefits/drawbacks are being considered and try to get as much information out in the open as possible.  From your earlier comments it sounds like there was internal discussion at Docker, Inc. about this that hasn't made its way onto GitHub yet.

stevvooe commented:
`cznic/ql` was suggested by @philips, who works for coreos. And to tell you the truth, this is problem that can suffer from analysis paralysis.

Do you have any experience with deploying go projects with sqlite? Do you have any examples that aren't a nightmare? Every time I've tried to use it with a Go project, it has been a nightmare.

dmcgowan commented:
I suggest we merge this then open an issue related to alternative implementations and differential analysis

samuelkarp commented on snapshot/storage/metastore.go:
```diff
@@ -0,0 +1,79 @@
+
+	// Remove removes a snapshot from the metastore. Cleanup is called
+	// during the transaction to allow the driver to make data unavailable
+	// before committing. Cleanup should run as fast as possible, and if
```
What's the reasoning for the "as fast as possible" part here?  Is it just to ensure that the transaction scope (in time) is limited?

dmcgowan replied:
Since this is done in a serialized transaction, other operations would be blocked by this. This is intended to allow for making the commit 2 phase (db and fs), the callback should then do the minimal amount to make the object unavailable but defer cleanup. Can also make this clearer in the documentation.

samuelkarp commented on snapshot/storage/metastore.go:
```diff
@@ -0,0 +1,79 @@
+	// before committing. Cleanup should run as fast as possible, and if
+	// cleanup cannot be performed quickly, it should record the ids to
+	// cleanup and call after this function returns without error.
+	Remove(ctx context.Context, key string, cleanup func(id string, k snapshot.Kind) error) error
```
In `CreateActive` and `Commit`, you've broken the callback function out into an `opts` struct.  Any particular reason for avoiding that here?

dmcgowan replied:
Broke them out more for cleanliness than consistency. Might break it out later for consistency, I couldn't justify a single item option struct though.

samuelkarp replied:
I would prefer consistency, but we can come back to this later.

dmcgowan replied:
Added `RemoveOpts`

samuelkarp commented on snapshot/storage/ql/ql.go:
```diff
@@ -0,0 +1,358 @@
+	ms := &qlMetastore{
+		dbfile: dbfile,
+	}
+	schema := []string{
```
Might be good to add a TODO about schema migrations in here.  We'll need to settle on that before this implementation really moves beyond a prototype.

samuelkarp commented on snapshot/storage/ql/ql.go:
```diff
@@ -0,0 +1,358 @@
+		dbfile: dbfile,
+	}
+	schema := []string{
+		`CREATE TABLE IF NOT EXISTS Snapshots (Name string NOT NULL, Parent int, Active bool, Readonly bool);`,
```
How come `Parent` is an int and not a string?

I'm wondering if it might be clearer to have `Active` instead be an enumerated type `SnapshotType`.

It'd likely also be a good idea to include a few more things here:
* Creation timestamp
* Update timestamp
* Deletion (toombstone) timestamp

(Toombstoning/soft deleting can be a good thing to build in from the start, and then handle hard deletion/garbage collection separately.  This will enable the database to be repaired/recovered by an operator if something goes sideways.)

dmcgowan replied:
`Parent` is an int so that it can be joined based on `id()`, which is int. The identifier given back to the snapshot driver is always a unique string, but that can always be created from the integer.

dmcgowan replied:
> It'd likely also be a good idea to include a few more things here:

reasonable, can defer that for now though until we settle on an implementation

dmcgowan replied:
> Deletion (tombstone) timestamp

Not a discussion we need to have now but possibly in the future. My immediate take is that we shouldn't need to mark deletions with a consistent backend. Having both a consistent database and data stored on disk does allow getting out of sync, making the commits multi-phase helps but a crash may still require cleaning up. I think we should focus on the database being consistent and if we have a cleanup process it would only need to cleanup leftover files on disk using the database as the source of truth.

samuelkarp replied:
> `Parent` is an int so that it can be joined based on `id()`, which is int.

Is `id()` defined somewhere explicit or is that implicit in the schema for `ql`?

>  My immediate take is that we shouldn't need to mark deletions with a consistent backend. Having both a consistent database and data stored on disk does allow getting out of sync, making the commits multi-phase helps but a crash may still require cleaning up. I think we should focus on the database being consistent and if we have a cleanup process it would only need to cleanup leftover files on disk using the database as the source of truth.

It's not just about being consistent, but also about ensuring we have the right set of primitives to make the system repairable if something goes wrong.  This is an extremely common mitigation for disaster scenarios and is also useful for investigation if bugs get introduced.

dmcgowan replied:
> Is id() defined somewhere explicit or is that implicit in the schema for ql?

implicit for `ql`, the primary key is not defined but always there

> This is an extremely common mitigation for disaster scenarios and is also useful for investigation if bugs get introduced.

Agreed we should re-address this later. I wouldn't want to add this kind of retention without a process in place to clean it up to avoid unbounded growth.

samuelkarp replied:
> Agreed we should re-address this later. I wouldn't want to add this kind of retention without a process in place to clean it up to avoid unbounded growth.

Sounds good.  Can you add a TODO for this?

dmcgowan commented on snapshot/storage/metastore.go:
```diff
@@ -0,0 +1,79 @@
+type CreateActiveOpts struct {
+	Parent     string
+	Readonly   bool
+	AllParents bool
```
I didn't include the btrfs implementation but I have it ready and probably should. `AllParents` specifies that all parent IDs should be returned in the `Active` object. `overlay` needs this, `btrfs` does not, no need to do extra lookups for `btrfs`. I will document this.

samuelkarp replied:
Is there a cost to including all the `ParentID`s regardless of whether the `Snapshotter` needs it?  (I'm wondering if the cognitive overhead of knowing whether you care about the parents is worth making the interface slightly more complicated.)

dmcgowan replied:
Got rid of `AllParents`, it is ql specific and cannot demonstrate it saves anything to justify the complexity.

dmcgowan commented:
Rebased to get tests passing

@samuelkarp clarified some of the meta store documentation. Removed mention of the cleanup method needing to be fast, rather focusing on its importance in making content unavailable but also restorable should the database commit not succeed.

samuelkarp commented on snapshot/storage/metastore.go:
```diff
@@ -0,0 +1,83 @@
+
+	// Commit renames the active snapshot transaction referenced by `key`
+	// as a committed snapshot referenced by `Name`. The resulting snapshot
+	// will be committed and readonly and the `key` reference will no longer
```
Why is `key` no longer available for lookup or removal?  Based on [the diagram](https://github.com/docker/containerd/blob/92b9262f632a6fb91988dd6a4db8ddca1c669743/design/snapshot_model.png) I would expect `key` to still be usable to reference the underlying `Active` snapshot and be available for further `Commit`s.

dmcgowan replied:
This implementation makes `Commit` terminal (matching the current overlay implementation). It does not match the model described in the `Snapshotter` definition but based on the conversations we had, the `Snapshotter` needs to change to add new method which just does `Snapshot` as a non-terminal action and the model accordingly.

samuelkarp replied:
Got it, thanks!  Seems like we need an update to the design doc and `Snapshotter` interface too (but not necessarily as part of this PR).

samuelkarp commented on snapshot/storage/metastore.go:
```diff
@@ -0,0 +1,83 @@
+	// only the immediate parent is included.
+	GetActive(ctx context.Context, key string, allParents bool) (Active, error)
+
+	// Remove removes a snapshot from the metastore. Cleanup is called
```
Does `Remove` cover both `Active` and `Committed` snapshots?

dmcgowan replied:
It does

samuelkarp commented on snapshot/storage/ql/ql.go:
```diff
@@ -0,0 +1,362 @@
+
+			}
+
+			if err := fn(ctx, info); err != nil {
```
Does `fn` inside `doDatabase` block anything?  It looks like this is executed while `ms.dbfile` is held open.  Should we read the results into a `[]snapshot.Info` and then call `fn` over each of them after `doDatabase` returns?

dmcgowan replied:
This I wasn't sure about, the walk could block other processes from being able to access the database file, but within the process transactions should be allowed to continue. I am not sure what the right answer is, obviously I thought blocking other processes from connecting was a fine trade off vs reading everything into an area and calling.

samuelkarp replied:
Would this block other goroutines in the same process?  What happens if in my `fn` I want to make changes (like `Commit` something)?

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "snapshot-overlay-metastore" git@github.com:dmcgowan/containerd.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353901680
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

dmcgowan commented:
Pushed up implementation with boltdb, currently it is an order of magnitude faster than ql based on the included benchmarks. Think we should consider either merging this PR with the understanding the metastore interface will change or close this and re-open when there is an implementation just based on boltdb.

samuelkarp commented:
> Think we should consider either merging this PR with the understanding the metastore interface will change [...]

This sounds good to me.  I won't be able to take a look at this PR again until next week, so if you're okay with more feedback after this has already been merged that's be fine with me.

dmcgowan commented:
@samuelkarp going to make an update for next week, going to create a transaction type and push it into the context to simplify the existing callback structure.

dmcgowan commented:
Will re-open after interface change
